include/engine/datafacade_provider.hpp

```diff
@@ -2,55 +2,85 @@
 #define OSRM_ENGINE_DATAFACADE_PROVIDER_HPP
 
 #include "engine/data_watchdog.hpp"
+#include "engine/datafacade.hpp"
 #include "engine/datafacade/contiguous_internalmem_datafacade.hpp"
 #include "engine/datafacade/process_memory_allocator.hpp"
 
 namespace osrm
 {
 namespace engine
 {
-
-template <typename AlgorithmT> class DataFacadeProvider
+namespace detail
 {
-    using FacadeT = datafacade::ContiguousInternalMemoryDataFacade<AlgorithmT>;
 
+template <typename AlgorithmT, template <typename A> class FacadeT> class DataFacadeProvider
+{
   public:
+    using Facade = FacadeT<AlgorithmT>;
+
     virtual ~DataFacadeProvider() = default;
 
-    virtual std::shared_ptr<const FacadeT> Get() const = 0;
+    virtual std::shared_ptr<const Facade> Get() const = 0;
 };
 
-template <typename AlgorithmT> class ImmutableProvider final : public DataFacadeProvider<AlgorithmT>
+template <typename AlgorithmT, template <typename A> class FacadeT>
+class ImmutableProvider final : public DataFacadeProvider<AlgorithmT, FacadeT>
 {
-    using FacadeT = datafacade::ContiguousInternalMemoryDataFacade<AlgorithmT>;
-
   public:
+    using Facade = typename DataFacadeProvider<AlgorithmT, FacadeT>::Facade;
+
     ImmutableProvider(const storage::StorageConfig &config)
-        : immutable_data_facade(std::make_shared<FacadeT>(
+        : immutable_data_facade(std::make_shared<Facade>(
               std::make_shared<datafacade::ProcessMemoryAllocator>(config)))
     {
     }
 
-    std::shared_ptr<const FacadeT> Get() const override final { return immutable_data_facade; }
+    std::shared_ptr<const Facade> Get() const override final { return immutable_data_facade; }
 
   private:
-    std::shared_ptr<const FacadeT> immutable_data_facade;
+    std::shared_ptr<const Facade> immutable_data_facade;
+};
+
+template <typename AlgorithmT, template <typename A> class FacadeT>
```

Comment:
it is better to have one templated method here as
```
template <typename AlgorithmT, template <typename A> class FacadeT>
class WatchingProvider : public DataFacadeProvider<AlgorithmT, FacadeT>
{
    DataWatchdog<AlgorithmT> watchdog;
    using Facade = typename DataFacadeProvider<AlgorithmT, FacadeT>::Facade;

  public:
    std::shared_ptr<const Facade> Get() const override final
    {
        return watchdog.Get();
    }
};
```

For some reason clang uses the primary template and gcc specialized, but it is better to remove the ambiguity